traits/src/ff/mod.rs

```diff
@@ -0,0 +1,194 @@
+    + for<'a> core::ops::Add<&'a Self, Output = Self>
+    + for<'a> core::ops::Sub<&'a Self, Output = Self>
+    + for<'a> core::ops::Mul<&'a Self, Output = Self>
+    + for<'a> core::iter::Sum<&'a Self>
+    + for<'a> core::iter::Product<&'a Self>
+    + core::ops::AddAssign
+    + core::ops::SubAssign
+    + core::ops::MulAssign
+    + for<'a> core::ops::AddAssign<&'a Self>
+    + for<'a> core::ops::SubAssign<&'a Self>
+    + for<'a> core::ops::MulAssign<&'a Self>
+    + From<u64>
+{
+    /// The zero element of the field, the additive identity.
+    const ZERO: Self;
+
+    /// The one element of the field, the multiplicative identity.
+    const ONE: Self;
+
+    /// Returns an element chosen uniformly at random using a user-provided RNG.
+    fn random(rng: impl RngCore) -> Self;
+
+    /// Returns true iff this element is zero.
+    fn is_zero(&self) -> subtle::Choice {
+        self.ct_eq(&Self::ZERO)
+    }
+
+    /// Interpret a string of decimal digits as a (congruent) field element.
+    fn from_decimal_str(s: &str) -> Option<Self> {
+        crate::ff::utils::from_str(s, 10)
+    }
+
+    /// Interpret a string of hex digits as a (congruent) field element.
+    fn from_hex_str(s: &str) -> Option<Self> {
+        crate::ff::utils::from_str(s, 16)
+    }
+
+    /// Squares this element.
+    #[must_use]
+    fn square(&self) -> Self {
+        *self * self
+    }
+
+    /// Doubles this element.
+    #[must_use]
+    fn double(&self) -> Self {
+        *self + self
+    }
+
+    /// Returns the square root of the field element, if it is
+    /// quadratic residue.
+    fn sqrt(&self) -> subtle::CtOption<Self>;
+
+    /// Computes the multiplicative inverse of this element,
+    /// failing if the element is zero.
+    fn invert(&self) -> subtle::CtOption<Self>;
+
+    /// Exponentiates `self` by `exp`, where `exp` is a little-endian order integer
+    /// exponent.
+    ///
+    /// # Guarantees
+    ///
+    /// This operation is constant time with respect to `self`, for all exponents with the
+    /// same number of digits (`exp.as_ref().len()`). It is variable time with respect to
+    /// the number of digits in the exponent.
+    fn pow<S: AsRef<[u64]>>(&self, exp: S) -> Self {
+        crate::ff::arithmetic::pow(self, exp)
+    }
+
+    /// Exponentiates `self` by `exp`, where `exp` is a little-endian order integer
+    /// exponent.
+    ///
+    /// # Guarantees
+    ///
+    /// **This operation is variable time with respect to `self`, for all exponent.** If
+    /// the exponent is fixed, this operation is effectively constant time. However, for
+    /// stronger constant-time guarantees, [`Field::pow`] should be used.
+    fn pow_vartime<S: AsRef<[u64]>>(&self, exp: S) -> Self {
+        crate::ff::arithmetic::pow_vartime(self, exp)
+    }
+
+    /// Returns a field element that is congruent to the provided little endian unsigned
+    /// byte representation of an integer.
+    fn from_uniform_bytes(bytes: &[u8]) -> Self;
```

Comment:
Could we also add `to_uniform_bytes`?  Something like this:
`fn to_uniform_bytes(&self, bytes: &mut [u8])` which writes the congruent value as little endian in `bytes`.

Reply:
There is no single result for `to_uniform_bytes` since `from_uniform_bytes` is kind of modulo operation against a larger number against the order

Reply:
> There is no single result for `to_uniform_bytes` since `from_uniform_bytes` is kind of modulo operation

Could you elaborate on this?  There should be a single little endian representation of the integer modulo p, right?

Reply:
we allow `from_uniform_bytes` to take a value larger than modulus. For example in halo2 `ff::FromUniformBytes<64>` is used to generate challenges from 64 byte hash value. So this value is apperently is larger than field and modulo operation is applied to map that hash output to field.

Reply:
Aaah, now I understand the name of this function.  It's like "build a field from a list of bytes sampled from a uniform distribution of the byte slice space".